plasmapy/simulation/particle_tracker.py

```diff
@@ -0,0 +1,842 @@
+    "ParticleTracker",
+    "TimeElapsedStopCondition",
+]
+
+import astropy.units as u
+import collections
+import h5py
+import numpy as np
+import sys
+import warnings
+
+from abc import ABC, abstractmethod
+from collections.abc import Iterable
+from pathlib import Path
+from tqdm import tqdm
+from typing import Optional, Union
+
+from plasmapy.particles import Particle, particle_input
+from plasmapy.plasma.grids import AbstractGrid
+from plasmapy.plasma.plasma_base import BasePlasma
+from plasmapy.simulation.particle_integrators import boris_push
+
+
+class AbstractStopCondition(ABC):
+    """Abstract base class containing the necessary methods for a ParticleTracker stopping condition."""
+
+    @property
+    def tracker(self):
+        """Return the `ParticleTracker` object for this stop condition."""
+        return self._particle_tracker
+
+    @tracker.setter
+    def tracker(self, particle_tracker):
+        self._particle_tracker = particle_tracker
+
+    @property
+    @abstractmethod
+    def require_uniform_dt(self):
+        """Return whether or not this stop condition requires a uniform dt to be specified."""
+        ...
+
+    @property
+    @abstractmethod
+    def progress_description(self):
+        """Return a small string describing the relevant quantity shown on the meter."""
+        ...
+
+    @property
+    @abstractmethod
+    def units(self):
+        """Return the units of `total`."""
+        ...
+
+    @property
+    @abstractmethod
+    def is_finished(self):
+        """Return `True` if the simulation has finished."""
+        ...
+
+    @property
+    @abstractmethod
+    def progress(self):
+        """Return a number representing the progress of the simulation (compared to total).
+        This number represents the numerator of the completion percentage.
+        """
+        ...
+
+    @property
+    @abstractmethod
+    def total(self):
+        """Return a number representing the total number of steps in a simulation.
+        This number represents the denominator of the completion percentage.
+        """
+        ...
+
+
+class TimeElapsedStopCondition(AbstractStopCondition):
+    """Stop condition corresponding to the elapsed time of a ParticleTracker."""
+
+    def __init__(self, stop_time: u.Quantity):
+        self._particle_tracker = None
+        self.stop_time = stop_time.to(u.s).value
+
+    @property
+    def require_uniform_dt(self):
+        """The elapsed time stop condition requires a uniform time step
+        to stop all particles at the same time.
+        """
+        return True
+
+    @property
+    def progress_description(self):
+        """The time elapsed stop condition depends on elapsed time,
+        therefore the relevant quantity is time remaining.
+        """
+        return "Time remaining"
+
+    @property
+    def units(self):
+        """The units for the time elapsed condition have the units of seconds."""
+
+        return "seconds"
```

Comment:
In PlasmaPy we should use astropy units to represent units for consistency, so this should be `u.s`. 

Then, when you initiate the progress bar, you can convert that to a string 

```python
pbar = tqdm(
            initial=0,
            total=stop_condition.total,
            disable=not self.verbose,
            desc=stop_condition.progress_description,
            unit=str(stop_condition.units),
            bar_format="{l_bar}{bar}{n:.1e}/{total:.1e} {unit}",  # noqa: FS003
            file=sys.stdout,
        )
```

That will end up printing `s` instead of `seconds`, but I think it's better to do that and remain consistent with the expectation that astropy. If you really wanted to have it print seconds, you could do something like 

```python
if len(stop_condition.units.long_names)>0:
    unit_name = stop_condition.units.long_names[0] + 's'
else:
    unit_name = str(stop_condition.units)
```

The `long_names` list contains any fully spelled out unit names associated with a unit object (not plural)

Reply:
By using `unit_string` am I able to still return "seconds"??

Reply:
Yep sure, I think that can be whatever string representation you want.